sys/btree/btree.c

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright (C) 2020 Inria
+ * Copyright (C) 2020 Koen Zandberg <koen@bergzand.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sys_btree2
+ * @{
+ * @file
+ * @brief   Function implementations for btree structures
+ *
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include "btree.h"
+
+static void _path_helper(btree_node_t *node, size_t depth, void *ctx)
+{
+    (void)node;
+    size_t *max_depth = ctx;
+    if (*max_depth < depth) {
+        *max_depth = depth;
+    }
+}
+
+static size_t _max_path(btree_node_t *node)
+{
+    size_t len = 0;
+    btree_t start = {.start = node};
+    btree_traverse(&start, _path_helper, &len);
+    return len;
+}
+
+/**
+ * @brief find the node with a matching key
+ *
+ * @param   node    parent node
+ */
+static btree_node_t *_find_key(btree_node_t *cur_node, btree_node_t **parent, uint32_t key)
+{
+    while (cur_node) {
+        if (cur_node->key > key) {
+            *parent = cur_node;
+            cur_node = cur_node->left;
+        }
+        else if (cur_node->key < key) {
+            *parent = cur_node;
+            cur_node = cur_node->right;
+        }
+        else {
+            break;
+        }
+    }
+    return cur_node;
+}
+
+static size_t _find_distance(btree_node_t *parent, btree_node_t *child)
+{
+    size_t depth = 0;
+    while (child && (parent != child)) {
+        depth++;
+        child = child->parent;
+    }
+    return depth;
+}
+
+static btree_node_t *_find_min(btree_node_t *node)
+{
+    btree_node_t *min = NULL;
+    _find_key(node, &min, 0);
+    return min;
+}
+
+static btree_node_t *_find_max(btree_node_t *node)
+{
+    btree_node_t *max = NULL;
+    _find_key(node, &max, UINT32_MAX);
+    return max;
+}
```

Comment:
Won't those return `NULL` most of the time?
`_find_key()` only returns exact matches.